lineax/_solver/lsmr.py

```diff
@@ -0,0 +1,583 @@
+
+        # Initialize variables for 1st iteration.
+        itn = 0
+        zetabar = alpha * beta
+        alphabar = alpha
+        rho = 1.0
+        rhobar = 1.0
+        cbar = 1.0
+        sbar = 0.0
+
+        loop_state_main = (itn, alpha, beta, zetabar, alphabar, rho, rhobar, cbar, sbar)
+
+        # Initialize variables for estimation of ||r||.
+        betadd = beta
+        betad = 0.0
+        rhodold = 1.0
+        tautildeold = 0.0
+        thetatilde = 0.0
+        zeta = 0.0
+        d = 0.0
+
+        loop_state_r_est = (betadd, betad, rhodold, tautildeold, thetatilde, zeta, d)
+
+        # Initialize variables for estimation of ||A|| and cond(A)
+        normA2 = alpha**2
+        maxrbar = 0.0
+        minrbar = 1e100
+        normA = jnp.sqrt(normA2)
+        condA = 1.0
+
+        loop_state_anorm = (normA2, maxrbar, minrbar, normA, condA)
+
+        # Items for use in stopping rules, normb set earlier
+        istop = 0
+        ctol = 0.0
+        if self.conlim > 0:
+            ctol = 1 / self.conlim
+        normr = beta
+
+        loop_state_stopping = (istop, normr, normb)
+
+        loop_state = (
+            loop_state_main,
+            loop_state_r_est,
+            loop_state_anorm,
+            loop_state_stopping,
+            loop_state_vecs,
+        )
+
+        def condfun(loop_state):
+            (
+                loop_state_main,
+                loop_state_r_est,
+                loop_state_anorm,
+                loop_state_stopping,
+                loop_state_vecs,
+            ) = loop_state
+            (
+                itn,
+                alpha,
+                beta,
+                zetabar,
+                alphabar,
+                rho,
+                rhobar,
+                cbar,
+                sbar,
+            ) = loop_state_main
+            (
+                betadd,
+                betad,
+                rhodold,
+                tautildeold,
+                thetatilde,
+                zeta,
+                d,
+            ) = loop_state_r_est
+            (normA2, maxrbar, minrbar, normA, condA) = loop_state_anorm
+            (istop, normr, normb) = loop_state_stopping
+            (x, u, v, h, hbar) = loop_state_vecs
+
+            # Compute norms for convergence testing.
+            normar = abs(zetabar)
+            normx = self.norm(x)
+
+            # Now use these norms to estimate certain other quantities,
+            # some of which will be small near a solution.
+            test1 = normr / normb
+            test2 = lax.select((normA * normr) != 0, normar / (normA * normr), jnp.inf)
+            test3 = 1 / condA
+            t1 = test1 / (1 + normA * normx / normb)
+            rtol = self.btol + self.atol * normA * normx / normb
+
+            # bitmask for istop for possibly multiple conditions?
+
+            # x is an approximate solution to A@x = b, according to atol and btol.
+            istop += (test1 <= rtol).astype(int) * 2**1
```

Comment:
Nit: in general, prefer an explicit `jnp....` dtype rather than a plain `int`, as this actually has different precisions between Windows and Unix.
(It doesn't really matter here of course, where the numbers are small.)